File: app/src/main/java/com/velaphi/untamed/features/animalDetails/models/AnimalDetailsModel.java
package com.velaphi.untamed.features.animalDetails.models;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;

import com.google.firebase.firestore.IgnoreExtraProperties;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.util.List;


@IgnoreExtraProperties
@Entity(tableName = "favorite_animals")
public class AnimalDetailsModel implements Parcelable {

    @TypeConverters(Converters.class)
    private Description description;

    @TypeConverters(Converters.class)
    private Habitat habitat;

    @ColumnInfo
    @SerializedName("image")
    @Expose
    private String image;

    @ColumnInfo(name = "image_list")
    @TypeConverters(Converters.class)
    private List<String> imageList = null;

    @ColumnInfo(name = "basic_Info")
    @TypeConverters(Converters.class)
    private List<Info> basicInfo = null;

    @ColumnInfo(name = "located")
    @TypeConverters(Converters.class)
    private List<String> located;

    @ColumnInfo
    @SerializedName("key")
    @Expose
    private String key;

    @ColumnInfo
    @SerializedName("level")
    @Expose
    private int level;

    @PrimaryKey
    @SerializedName("name")
    @Expose
    @NonNull
    private String name;

    @ColumnInfo(name = "challenges")
    @TypeConverters(Converters.class)
    private List<Challenge> challenges = null;

    @ColumnInfo(name = "predators")
    @TypeConverters(Converters.class)
    private List<String> predators = null;

    @ColumnInfo(name = "scientific_names")
    @TypeConverters(Converters.class)
    private List<String> scientificNames = null;

    @ColumnInfo(name = "weight")
    @TypeConverters(Converters.class)
    private List<String> weight = null;

    @SerializedName("quotable")
    @Expose
    private String quotable;

    @ColumnInfo(name = "facts")
    @TypeConverters(Converters.class)
    private List<Fact> facts = null;

    @ColumnInfo(name = "video_list")
    @TypeConverters(Converters.class)
    private List<Video> videoList = null;

    public static final Creator<AnimalDetailsModel> CREATOR = new Creator<AnimalDetailsModel>() {
        @Override
        public AnimalDetailsModel createFromParcel(Parcel in) {
            return new AnimalDetailsModel(in);
        }

        @Override
        public AnimalDetailsModel[] newArray(int size) {
            return new AnimalDetailsModel[size];
        }
    };

    protected AnimalDetailsModel(Parcel in) {
        description = in.readParcelable(Description.class.getClassLoader());
        habitat = in.readParcelable(Habitat.class.getClassLoader());
        image = in.readString();
        imageList = in.createStringArrayList();
        basicInfo = in.createTypedArrayList(Info.CREATOR);
        located = in.createStringArrayList();
        key = in.readString();
        level = in.readInt();
        name = in.readString();
        challenges = in.createTypedArrayList(Challenge.CREATOR);
        predators = in.createStringArrayList();
        scientificNames = in.createStringArrayList();
        videoList = in.createTypedArrayList(Video.CREATOR);
        weight = in.createStringArrayList();
        quotable = in.readString();
        facts = in.createTypedArrayList(Fact.CREATOR);
    }


    public AnimalDetailsModel() {
    }

    public static Creator<AnimalDetailsModel> getCREATOR() {
        return CREATOR;
    }

    public List<Video> getVideoList() {
        return videoList;
    }

    public Description getDescription() {
        return description;
    }


    public Habitat getHabitat() {
        return habitat;
    }

    public String getImage() {
        return image;
    }

    public List<String> getImageList() {
        return imageList;
    }

    public String getKey() {
        return key;
    }

    public int getLevel() {
        return level;
    }

    public List<Info> getBasicInfo() {
        return basicInfo;
    }

    public List<String> getLocated() {
        return located;
    }

    public List<String> getPredators() {
        return predators;
    }

    public List<String> getScientificNames() {
        return scientificNames;
    }

    public List<String> getWeight() {
        return weight;
    }

    public String getQuotable() {
        return quotable;
    }

    public List<Fact> getFacts() {
        return facts;
    }

    public String getName() {
        return name;
    }

    public List<Challenge> getChallenges() {
        return challenges;
    }

    public void setDescription(Description description) {
        this.description = description;
    }

    public void setHabitat(Habitat habitat) {
        this.habitat = habitat;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public void setImageList(List<String> imageList) {
        this.imageList = imageList;
    }

    public void setBasicInfo(List<Info> basicInfo) {
        this.basicInfo = basicInfo;
    }

    public void setLocated(List<String> located) {
        this.located = located;
    }

    public void setKey(String key) {
        this.key = key;
    }

    public void setLevel(int level) {
        this.level = level;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setChallenges(List<Challenge> challenges) {
        this.challenges = challenges;
    }

    public void setPredators(List<String> predators) {
        this.predators = predators;
    }

    public void setScientificNames(List<String> scientificNames) {
        this.scientificNames = scientificNames;
    }

    public void setWeight(List<String> weight) {
        this.weight = weight;
    }

    public void setQuotable(String quotable) {
        this.quotable = quotable;
    }

    public void setFacts(List<Fact> facts) {
        this.facts = facts;
    }

    public void setVideoList(List<Video> videoList) {
        this.videoList = videoList;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(description, flags);
        dest.writeParcelable(habitat, flags);
        dest.writeString(image);
        dest.writeStringList(imageList);
        dest.writeTypedList(basicInfo);
        dest.writeStringList(located);
        dest.writeString(key);
        dest.writeInt(level);
        dest.writeString(name);
        dest.writeTypedList(challenges);
        dest.writeStringList(predators);
        dest.writeStringList(scientificNames);
        dest.writeTypedList(videoList);
        dest.writeStringList(weight);
        dest.writeString(quotable);
        dest.writeTypedList(facts);
    }
}

File: app/src/main/java/com/velaphi/untamed/features/old_animalDetails/models/AnimalDetailsModel.kt
package com.velaphi.untamed.features.old_animalDetails.models

import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.NonNull
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.TypeConverters
import com.google.firebase.firestore.IgnoreExtraProperties
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import java.util.*

@IgnoreExtraProperties
@Entity(tableName = "favorite_animals")
class AnimalDetailsModel : Parcelable {
    @TypeConverters(Converters::class)
    var description: Description? = null
    @TypeConverters(Converters::class)
    var habitat: Habitat? = null
    @ColumnInfo
    @SerializedName("image")
    @Expose
    var image: String? = null
    @ColumnInfo(name = "image_list")
    @TypeConverters(Converters::class)
    var imageList: List<String>? = null
    @ColumnInfo(name = "basic_Info")
    @TypeConverters(Converters::class)
    var basicInfo: List<Info>? = null
    @ColumnInfo(name = "located")
    @TypeConverters(Converters::class)
    var located: List<Country>? = null
    @ColumnInfo
    @SerializedName("key")
    @Expose
    var key: String? = null
    @ColumnInfo
    @SerializedName("level")
    @Expose
    var level = 0
    @PrimaryKey
    @SerializedName("name")
    @Expose
    @NonNull
    var name: String? = null
    @ColumnInfo(name = "challenges")
    @TypeConverters(Converters::class)
    var challenges: List<Challenge>? = null
    @ColumnInfo(name = "predators")
    @TypeConverters(Converters::class)
    var predators: List<Predator>? = null
    @ColumnInfo(name = "prey")
    @TypeConverters(Converters::class)
    var prey: List<Prey>? = null
    @ColumnInfo(name = "scientific_names")
    @TypeConverters(Converters::class)
    var scientificNames: List<String>? = null
    @ColumnInfo(name = "weight")
    @TypeConverters(Converters::class)
    var weight: List<String>? = null
    @SerializedName("quotable")
    @Expose
    var quotable: String? = null
    @ColumnInfo(name = "facts")
    @TypeConverters(Converters::class)
    var facts: List<Fact>? = null
    @ColumnInfo(name = "video_list")
    @TypeConverters(Converters::class)
    var videoList: ArrayList<Video>? = null

    constructor(parcel: Parcel) : this() {
        description = parcel.readParcelable(Description::class.java.classLoader)
        habitat = parcel.readParcelable(Habitat::class.java.classLoader)
        image = parcel.readString()
        imageList = parcel.createStringArrayList()
        basicInfo = parcel.createTypedArrayList(Info)
        located = parcel.createTypedArrayList(Country)
        key = parcel.readString()
        level = parcel.readInt()
        name = parcel.readString()
        challenges = parcel.createTypedArrayList(Challenge)
        predators = parcel.createTypedArrayList(Predator)
        prey = parcel.createTypedArrayList(Prey)
        scientificNames = parcel.createStringArrayList()
        weight = parcel.createStringArrayList()
        quotable = parcel.readString()
        facts = parcel.createTypedArrayList(Fact)
    }

    constructor() {}


    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeParcelable(description, flags)
        parcel.writeParcelable(habitat, flags)
        parcel.writeString(image)
        parcel.writeStringList(imageList)
        parcel.writeTypedList(basicInfo)
        parcel.writeTypedList(located)
        parcel.writeString(key)
        parcel.writeInt(level)
        parcel.writeString(name)
        parcel.writeTypedList(challenges)
        parcel.writeTypedList(predators)
        parcel.writeTypedList(prey)
        parcel.writeStringList(scientificNames)
        parcel.writeStringList(weight)
        parcel.writeString(quotable)
        parcel.writeTypedList(facts)
    }

    companion object CREATOR : Parcelable.Creator<AnimalDetailsModel> {
        override fun createFromParcel(parcel: Parcel): AnimalDetailsModel {
            return AnimalDetailsModel(parcel)
        }

        override fun newArray(size: Int): Array<AnimalDetailsModel?> {
            return arrayOfNulls(size)
        }
    }

}